Reworks trashbags slightly

LemonInTheDark opened the pull request:
## About The Pull Request

I'm a bit sad about the state of trashbags. 
They're very clunky to use, so they almost never get touched. S depressing. Let's try and fix that.

Let's make em fit in the belt slot (again), but as a tradeoff we'll make it harder to pull one thing from your bag.
We'll give it a say, 1.5 second delay, so you can't quickdraw from em.
If you try and dump them out into something else, we'll throw any spillover on the ground below you

I'm also doing some general code cleanup here. Making procs more readable, vars more direct, removing some old legacy stuff.
I've added a remove_single proc to hook into via subtype, which takes a mob as input. this has required placing extra requirement on some helper procs, but fortunately it's not something they're unable to meet.

My hope is this will make garbage bags usable without being stupid.

## Why It's Good For The Game

I don't see these get used at all, cause they're a pain to carry around. They got gimped because people were using them as infinite storage for shotgun shells and other small items. 
I've made using them for this sort of thing hard and slow, so I think we oughta be fine. If not I'll do some more touching, maybe give the autodrop a delay.

## Changelog
:cl:
balance: The janitor's trashbag now fits on his belt. In exchange, taking something out of it sends a visible message, and has a delay.
/:cl:

LemonInTheDark committed:
Reworks trashbags slightly

I'm a bit sad about the state of trashbags. They're very clunky to use,
so they almost never get touched. S sad. Let's try and fix that.

Let's make em fit in the belt slot (again), but as a tradeoff we'll make
it impossible to pull just one thing out of the bag. Emptying it is an
all at once affair, and you can't peek inside them either.

If you try and dump them out into something else, we'll throw any
spillover on the ground below you

I'm also doing some general code cleanup here. Making procs more
readable, vars more direct, removing some old legacy stuff.

My hope is this will make garbage bags usable without being stupid

AlamoTurtle commented:
I mean, people do root through trash bags sometimes, though I'm unsure how difficult it'd be to add a delay to opening a bag to see inside it. At least in hand, old behavior might be better off for anyone who genuinely just wants to dig through trash for a rat or whatever, with it not mounted to a belt slot. Maybe a movement debuff based on how much crap is in it, so a full bag slows you down about as much as a duffel? Just tossing stuff out there, because it sounds like a pain in the ass for dumpster divers looking for a bungo pit to have to spew trash everywhere.

Love the idea of giving those stupid bags some use tho.

VladinXXV commented:
Trashbags can pick up a lot of different things. And sometimes those things do not belong in the trash. Rather than "you have to dump out and alt-click/right-click thus crashing your game", do you think you could put a delay on opening the bag **_and_** 
 on manually removing items? This would make it a very poor choice of bandolier, but is a minor inconvenience to the janitor who isn't trying to throw important things away (or is trying to eat the rats they found).

Other ideas to consider:
- instead of taking one thing out, you will take out all of a **_type_** of thing.
- I would really like to see the trashbag fit in the suit storage of the janitor biosuit and janitor winter coat. I think fitting in this slot is better than going on the belt, because the belt slot is for the janibelt. If you make it fit in these storage slots INSTEAD of the belt, it also makes it less appealing as a powergame storage due to restricting the use of armor.

LemonInTheDark commented:
> Trashbags can pick up a lot of different things. And sometimes those things do not belong in the trash. Rather than "you have to dump out and alt-click/right-click thus crashing your game", do you think you could put a delay on opening the bag **_and_** on manually removing items? This would make it a very poor choice of bandolier, but is a minor inconvenience to the janitor who isn't trying to throw important things away (or is trying to eat the rats they found).

Alt or right clicking 30/60 (more realistically 15-30) items should not crash your game. I'd be in theory ok with adding a delay to opening the bag/selecting items, it's just more troublesome because of how direct storage code is. I'll consider it tho.

> Other ideas to consider:
> 
>     * when removing items manually, instead of taking one thing out, you will take out all of a **_type_** of thing.

This would be kinda weird to do, don't really think it should be done. Would feel wrong.

>     * I would really like to see the trashbag fit in the suit storage of the janitor biosuit and janitor winter coat. I think fitting in this slot is better than going on the belt, because the belt slot is for the janibelt. If you make it fit in these storage slots INSTEAD of the belt, it also makes it less appealing as a powergame storage due to restricting the use of armor.

I don't think removing the availability of armor would be enough to make it not a powergaming strategy. I like the choice you need to make between belt and trashbag, (at least in theory, not sure how it'll work out) and I don't want to undermine that or require a winter coat to be "fully kitted"

> I mean, people do root through trash bags sometimes, though I'm unsure how difficult it'd be to add a delay to opening a bag to see inside it. At least in hand, old behavior might be better off for anyone who genuinely just wants to dig through trash for a rat or whatever, with it not mounted to a belt slot. Maybe a movement debuff based on how much crap is in it, so a full bag slows you down about as much as a duffel? Just tossing stuff out there, because it sounds like a pain in the ass for dumpster divers looking for a bungo pit to have to spew trash everywhere.

Slowdown maybe if it's real full, though that might just make it useless, and I don't want that. I will think about making it pick throughable

Helg2 commented:
> Alt or right clicking 30/60 (more realistically 15-30) items should not crash your game. I'd be in theory ok with adding a delay to opening the bag/selecting items, it's just more troublesome because of how direct storage code is. I'll consider it tho.

Alt+Clicking 30 items on the floor is absolute cancer. No one likes it, and if its something like glass shards which make noise on the drop, it's cancer x2. Just remember that some people will just spam that shit, until everyone's ears will bleed out.

coiax commented:
If we are to go down the "takes time to root through, and takes time to remove" solution, we should make it generic, since I think lots of people would be interested in that for restrictions with storage.

san7890 commented on code/datums/storage/storage.dm:
```diff
@@ -952,32 +959,31 @@ GLOBAL_LIST_EMPTY(cached_storage_typecaches)
+		attempt_remove(to_remove)
+		INVOKE_ASYNC(src, PROC_REF(put_in_hands_async), to_show, to_remove)
+		if(!silent)
+			to_show.visible_message(span_warning("[to_show] draws [to_remove] from [resolve_parent]!"), span_notice("You draw [to_remove] from [resolve_parent]."))
```
```suggestion
			to_show.visible_message(
				span_warning("[to_show] draws [to_remove] from [resolve_parent]!"),
				span_notice("You draw [to_remove] from [resolve_parent]."),
			)
```

multiline to make it a bit nicer since it's a bit long imo

san7890 commented on code/game/objects/items/storage/bags.dm:
```diff
@@ -49,6 +48,22 @@
+
+	// Otherwise, we're gonna dump into the dest object
+	var/turf/dump_onto = get_turf(dest_object)
+	user.visible_message("[user] dumps the contents of [src] all out on \the [dump_onto]", "The remaining trash in \the [src] falls out onto \the [dump_onto]")
```
```suggestion
	user.visible_message(
		"[user] dumps the contents of [src] all out on \the [dump_onto]",
		"The remaining trash in \the [src] falls out onto \the [dump_onto]",
	)
```

LemonInTheDark replied:
yeah sure

AlamoTurtle commented:
> If we are to go down the "takes time to root through, and takes time to remove" solution, we should make it generic, since I think lots of people would be interested in that for restrictions with storage.

Nah. Backpacks have different zippers. Satchels have dividers and pockets. Many things have excuses for not being hard to sort through. A garbage bag is a garbage bag. Nonetheless, there has to be some sort of countermeasure that can prevent people using garbage bag belts, instead of having to dump 30 items on the ground and root through a poor alt-click/shift rightclick menu to find what they need or want.

Either way, trash bag changes good.

san7890 commented on code/game/objects/items/storage/bags.dm:
```diff
@@ -39,7 +39,6 @@
 	inhand_icon_state = "trashbag"
 	lefthand_file = 'icons/mob/inhands/equipment/custodial_lefthand.dmi'
 	righthand_file = 'icons/mob/inhands/equipment/custodial_righthand.dmi'
-	slot_flags = null
```
is CI failing because this line was removed? that's odd. i could be wrong though but that seems to be the only thing that could be causing the missing icon state? unsure.

LemonInTheDark replied:
I think it's cause they don't have a worn icon state, which do be depressing

LemonInTheDark replied:
handled

LemonInTheDark committed:
San's multiline

Co-authored-by: san7890 <the@san7890.com>

LemonInTheDark committed:
Adds a hook for removing a single item, refactors some procs to support it

Moves trashbags from being unable to view inside them to being unable to
draw from them without a delay

LemonInTheDark committed:
Merge branch 'garbage-DAY-SONABITCH' of https://github.com/LemonInTheDark/tgstation into garbage-DAY-SONABITCH

LemonInTheDark committed:
some extra spans

LemonInTheDark commented:
I've gone with the timed delay on removal idea. Provides a lever to touch if players fall back on old bad behavior.

LemonInTheDark committed:
whoops forgot the subtype

LemonInTheDark committed:
adds worn icons
